### PR TITLE
miscellaneous fixes

### PR DESCRIPTION
**Pull request by Dudemanguy**
Just some relatively low hanging fruit from the issue tracker. Could use some more testing on the second commit since nearly all of the subtitles I have are ASS.

### CODE REVIEW

**Comment by Flo56958** [comment]
I have done a bit of testing of the second commit. Tested 2 scenes I know that flicker and where subtitles end prematurely. These scenes work perfectly now for me with this fix. No flickering, no premature cutoff (if there are two subtitles on the screen for music lyrics and voices that switch lines independently). Subtitles are shown in these scenes as they should now.

**Comment by Dudemanguy** [comment]
Thanks, I suppose it should work for pgs then. I'm unsure if possibly some other subtitle format could regress, but I don't really even know what formats actually wind up using the sd_lavc codepath. Most things will be sd_ass and --sub-demuxer didn't seem to do anything but maybe I'm just doing it wrong.

File: player/command.c [hunk 1]
```diff
@@ -6954,8 +6954,12 @@ void mp_option_change_callback(void *ctx, struct m_config_option *co, int flags,
         }
     }
 
-    if (opt_ptr == &opts->pause)
+    if (opt_ptr == &opts->pause) {
+        // If we are going to pause, always drain the ao on EOF if it exists.
+        if (mpctx->ao && mpctx->audio_status == STATUS_EOF && opts->pause)
+            ao_drain(mpctx->ao);
```
Comment (sfan5) [inline player/command.c:6963]: hmm isn't there a better way to fit this into the audio state machine/code?
calling this from here smells like breaking abstraction

File: player/sub.c [hunk 2]
```diff
@@ -201,8 +201,10 @@ void reinit_sub(struct MPContext *mpctx, struct track *track)
     osd_set_sub(mpctx->osd, order, track->d_sub);
     sub_control(track->d_sub, SD_CTRL_SET_TOP, &order);
 
+    // When paused we have to wait for packets to be available.
+    // So just retry until we get a packet in this case.
     if (mpctx->playback_initialized)
-        update_subtitles(mpctx, mpctx->playback_pts);
+        while (!update_subtitles(mpctx, mpctx->playback_pts) && mpctx->paused);
```
[the remark this answers is not in the document]
Comment (Dudemanguy) [reply player/sub.c:207]: I'm trusting the the comments for update_subtitles which claims "false if the player should wait for new demuxer data, and then should retry". Could add an arbitrary timeout if you prefer? Not sure how long would be good.

File: player/command.c (hunk 1 above)
Comment (Dudemanguy) [reply player/command.c:6963]: Yes, quite an obvious place in retrospect. Thanks.

File: player/sub.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (sfan5) [reply player/sub.c:207]: Nah that's fine.

File: video/out/x11_common.c [hunk 3]
```diff
@@ -1579,10 +1579,10 @@ static void vo_x11_create_window(struct vo *vo, XVisualInfo *vis,
     if (!x11->parent) {
         vo_x11_update_composition_hint(vo);
         vo_x11_set_wm_icon(x11);
-        vo_x11_update_window_title(vo);
         vo_x11_dnd_init_window(vo);
         vo_x11_set_property_utf8(vo, XA(x11, _GTK_THEME_VARIANT), "dark");
     }
+    vo_x11_update_window_title(vo);
```
[the remark this answers is not in the document]
Comment (sfan5) [reply video/out/x11_common.c:1585]: sure but won't mpv and the other application step on each others toes when both update the window title?

File: audio/out/buffer.c [hunk 4]
```diff
@@ -355,6 +355,11 @@ void ao_set_paused(struct ao *ao, bool paused)
     bool wakeup = false;
     bool do_reset = false, do_start = false;
 
+    // If we are going to pause and ao is still playing,
+    // be sure to drain the ao first for gapless.
+    if (paused && ao_is_playing(ao))
+        ao_drain(ao);
```
Comment (sfan5) [inline audio/out/buffer.c:361]: hmm, I'm wondering if draining on every pause isn't too excessive or has negative side-effects.
non-gapless playback doesn't need this, right?

File: video/out/x11_common.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (Dudemanguy) [reply video/out/x11_common.c:1585]: Yeah they would have to distinguish between their own updates and ones from mpv. Maybe just make this an option instead (personally inclined to default it to yes)?

File: audio/out/buffer.c (hunk 4 above)
Comment (Dudemanguy) [reply audio/out/buffer.c:361]: For non-gapless, ao_is_playing is false. But this should only need to run on the end of file not every single pause. Let me see if there's a way to do that without having to modify ao_set_paused itself.

File: video/out/x11_common.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (Dudemanguy) [reply video/out/x11_common.c]: Dropped the x11 commit since it turned into a feature instead of a fix (it's #12132 now).

**Comment by Dudemanguy** [comment]
I just realized that the first commit's fix is incomplete. When going from an external subtitle to an internal one while paused, it still doesn't show. Note that the other way around (internal to external) works because of e928bd5fdb4e2a7cac3e55c4a2056c0cb4a76d9c. I'll look into this some more.

**Comment by Dudemanguy** [comment]
Sorry ignore the above I'm stupid. I was on the wrong branch when checking that. It works fine.

**Review by sfan5 — APPROVED** [review]

**Pull request merged**